src/resmom/mom_hook_func.c

```diff
@@ -3075,6 +3080,343 @@ record_job_last_hook_executed(unsigned int hook_event,
+				* must be all down
+				*/
+				req_reject(PBSE_SISCOMM, 0, preq); /* all sis down */
+			} else {
+				reply_ack(preq);	/* no sisters, reply now  */
+			}
+#endif
+		} else {
+			void post_reply(job *, int);
+			post_reply(pjob, 0);
+			mom_deljob(pjob);
+		}
+		free(hook_input);
+		return;
+	}
+
+	if (reboot_flag) {
+		if (phook->user == HOOK_PBSUSER) {
+			log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_HOOK,
+				LOG_INFO, phook->hook_name,
+				"Not allowed to issue reboot if run as user");
+		} else {
+			log_event(PBSEVENT_DEBUG, PBS_EVENTCLASS_HOOK,
+				LOG_INFO, phook->hook_name,
+				"requested for host to be rebooted");
+			do_reboot(reboot_cmd);
+		}
+	}
+
+	/* Create task to check and run next hook script */
+	new_task = set_task(WORK_Immed, 0,
+					(void*)run_execjob_end_hooks, hook_input);
+	if (!new_task) 
+		log_err(errno, __func__, 
+			"Unable to set task for run_execjob_end_hooks");
+	else
+		new_task->wt_parm2 = phook;
+
+	return;
+}
+
+/**
+ * @brief
+ * This function loops through the execjob_end hook list,
+ * and runs hook in the background.
+ * 
+ * @param[in] ptask - the work task. 
+ * 
+ * @retval 0 running a hook script.
+ * @retval 1 no hook script to run.
+ */
+ static int
+ run_execjob_end_hooks(struct work_task *ptask)
+ {
+	char	hook_infile[MAXPATHLEN+1] = {'\0'};
+	char	hook_outfile[MAXPATHLEN+1] = {'\0'};
+	char	hook_datafile[MAXPATHLEN+1] = {'\0'};
+	char	jobid[PBS_MAXSVRJOBID+1];
+	int		n;
+	int		ret;
+#if !MOM_ALPS
+	int 	numnodes = 0;
+	struct 	batch_request *preq;
+#endif
+
+	job              *pjob = NULL;
+	pbs_list_head    *head_ptr = NULL;
+	hook             *phook = NULL;
+	mom_hook_input_t *hook_input = (mom_hook_input_t *) ptask->wt_parm1;
+
+	if (ptask->wt_parm2 == NULL) {
+		head_ptr = &svr_execjob_end_hooks;
+		phook = (hook *)GET_NEXT(*head_ptr);
+	} else {
+		phook = (hook *)ptask->wt_parm2;
+		phook = (hook *)GET_NEXT(phook->hi_execjob_end_hooks);
+	}
+
+	while (phook) {
+		if (phook->enabled == FALSE) {
+			phook = (hook *)GET_NEXT(phook->hi_execjob_end_hooks);
+			continue;
+		}
+		if (phook->script == NULL) {
+			log_event(PBSEVENT_DEBUG3, PBS_EVENTCLASS_HOOK,
+			LOG_ERR, phook->hook_name,
+				"Hook has no script content. Skipping hook.");
+			phook = (hook *)GET_NEXT(phook->hi_execjob_end_hooks);
+			continue;
+		}
+		break;
+	}
+	if (!phook) {
+		if (ptask->wt_parm2 != NULL) {
+			pjob = hook_input->pjob;
+
+			if (pjob->ji_execjob_end_hook_event_started == IS_DISCARD_JOB) {
+
+				n = pjob->ji_wattr[(int)JOB_ATR_run_version].at_val.at_long;
+				strcpy(jobid, pjob->ji_qs.ji_jobid);
+				mom_deljob(pjob);
+
+				if ((ret=is_compose(server_stream, IS_DISCARD_DONE)) != DIS_SUCCESS)
+					goto err;
+
+				if ((ret=diswst(server_stream, jobid)) != DIS_SUCCESS)
+					goto err;
+
+				if ((ret=diswsi(server_stream, n)) != DIS_SUCCESS)
+					goto err;
+
+				rpp_flush(server_stream);
+				rpp_eom(server_stream); 
+
+			} else if (pjob->ji_qs.ji_svrflags & JOB_SVFLG_HERE) {
+#if MOM_ALPS
```

Comment:
This code is now triplicated. here, 3229, and in req_deletejob. Maybe move it to a function?